### ISSUE TITLE
Numba not vectorizing 2d copy loop

### ISSUE THREAD

**Issue by aseyboldt** [numba/numba#8314]
```python
import numpy as np
import numba
import llvmlite.binding as llvm

llvm.set_option('', '--debug-only=loop-vectorize,iv-descriptors')

@numba.njit("void(float64[:, ::1], float64[:, ::1])", fastmath=True, boundscheck=False)
def foo(a, out):
    n, m = a.shape
    assert out.shape == (n, m)
    for i in range(n):
        for j in range(m):
            out[i, j] = a[i, j]
```

This reports that the loop can not be vectorized, due to "an unidentified PHI".

```
LV: Checking a loop in "_ZN8__main__3fooB3v11B54c8tJTIcFHzwl2ILiXkcBV0KBSsOcbovu9mp1kJR6rSYw_2bIogqjUBE5ArrayIdLi2E1C7mutable7alignedE5ArrayIdLi2E1C7mutable7alignedE" from foo
LV: Loop hints: force=? width=0 unroll=0
LV: Found a loop: B52.us
LV: PHI is not a poly recurrence.
LV: PHI is not a poly recurrence.
LV: Not vectorizing: Found an unidentified PHI   %.319.sroa.0.021.us = phi i64 [ 0, %B40.endif.us ], [ %.319.sroa.0.0.us, %B52.us ]
LV: Interleaving disabled by the pass manager
LV: Can't vectorize the instructions or CFG
LV: Not vectorizing: Cannot prove legality.
LV: PHI is not a poly recurrence.
LV: PHI is not a poly recurrence.
```

This is using the development version of numba and the llvmlite package from the numba dev channel:

```
icc_rt                    2020.2                intel_254    numba/label/dev
llvmlite                  0.40.0dev0       py39he1b5a44_9    numba/label/dev
numba                     0.57.0.dev0+170.gce9bdacc7           dev_0    <develop>
```

**Comment by esc** [comment]
@aseyboldt thank you, I can confirm I am able to reproduce the LV output.

**Comment by stuartarchibald** [comment]
Debugging with Numba modified to use the new remarks capturing interface...
```
--- !Analysis
Pass:            loop-vectorize
Name:            NonReductionValueUsedOutsideLoop
DebugLoc:        { File: bug.py, Line: 13, Column: 1 }
Function:        _ZN8__main__3foo<mangling>
Args:
  - String:          'loop not vectorized: '
  - String:          value that could not be identified as reduction is used outside the loop
...
--- !Analysis
Pass:            loop-vectorize
Name:            CantIdentifyArrayBounds
DebugLoc:        { File: bug.py, Line: 13, Column: 1 }
Function:        _ZN8__main__3foo<mangling>
Args:
  - String:          'loop not vectorized: '
  - String:          cannot identify array bounds
...
--- !Missed
Pass:            loop-vectorize
Name:            MissedDetails
DebugLoc:        { File: bug.py, Line: 13, Column: 1 }
Function:        _ZN8__main__3foo<mangling>
Args:
  - String:          loop not vectorized
```
in the above, line 13 is the inner loop `for j in range(m)`. Two observations:
1. If the inner loop is changed to a `while` loop, the vectorizer works as expected. 
2. Running loop rotation at the end of the "cheap" optimisation pass also makes the inner loop vectorize:
   ```diff
   diff --git a/numba/core/codegen.py b/numba/core/codegen.py
   --- a/numba/core/codegen.py
   +++ b/numba/core/codegen.py
   @@ -1223,6 +1223,7 @@ class CPUCodegen(Codegen):
                # slightly higher cost:
                pm.add_licm_pass()
                pm.add_cfg_simplification_pass()
   +            pm.add_loop_rotate_pass()
            if config.LLVM_REFPRUNE_PASS:
                pm.add_refprune_pass(_parse_refprune_flags())
            return pm
   ```

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/9428 fixed this. With it, the OP's code prints:

```
LV: Checking a loop in "_ZN8__main__3fooB2v1B60c8tJTIcFKzyF2ILShI4CrgQElYZ5yRbdT9XqICn1Wk1gsBZBVGsCAA_3d_3dE5ArrayIdLi2E1C7mutable7alignedE5ArrayIdLi2E1C7mutable7alignedE" from foo
LV: Loop hints: force=? width=0 interleave=0
LV: Found a loop: B52.us
LV: Found an induction variable.
LV: We can vectorize this loop (with a runtime bound check)!
LV: Found trip count: 0
LV: The max safe fixed VF is: 33554432.
LV: The max safe scalable VF is: vscale x 0.
LV: The Smallest and Widest types: 64 / 64 bits.
LV: The Widest register safe to use is: 128 bits.
LV: The Widest register safe to use is: vscale x 0 bits.
LV: The target has no scalable vector registers.
LV: Found uniform instruction:   %exitcond.not = icmp eq i64 %.349.us, %arg.out.5.1
LV: Found uniform instruction:   %.448.us = getelementptr double, double* %arg.a.4, i64 %.447.us
LV: Found uniform instruction:   %.503.us = getelementptr double, double* %arg.out.4, i64 %.447.us
LV: Found uniform instruction:   %.447.us = add i64 %.298.07.us, %.444.us
LV: Found uniform instruction:   %.447.us = add i64 %.298.07.us, %.444.us
LV: Found uniform instruction:   %.298.07.us = phi i64 [ 0, %B40.endif.us ], [ %.349.us, %B52.us ]
LV: Found uniform instruction:   %.349.us = add nuw nsw i64 %.298.07.us, 1
LV: Found scalar instruction:   %.298.07.us = phi i64 [ 0, %B40.endif.us ], [ %.349.us, %B52.us ]
LV: Found scalar instruction:   %.349.us = add nuw nsw i64 %.298.07.us, 1
LV: Scalarizing:  %.447.us = add i64 %.298.07.us, %.444.us
LV: Scalarizing:  %.448.us = getelementptr double, double* %arg.a.4, i64 %.447.us
LV: Scalarizing:  %.449.us = load double, double* %.448.us, align 8
LV: Scalarizing:  %.503.us = getelementptr double, double* %arg.out.4, i64 %.447.us
LV: Scalarizing:  store double %.449.us, double* %.503.us, align 8
LV: Scalarizing:  %.447.us = add i64 %.298.07.us, %.444.us
LV: Scalarizing:  %.448.us = getelementptr double, double* %arg.a.4, i64 %.447.us
LV: Scalarizing:  %.503.us = getelementptr double, double* %arg.out.4, i64 %.447.us
VPlan 'Initial VPlan for VF={1},UF>=1' {
Live-in vp<%0> = vector-trip-count

<x1> vector loop: {
  B52.us:
    EMIT vp<%1> = CANONICAL-INDUCTION
    WIDEN-INDUCTION %.298.07.us = phi 0, %.349.us
    CLONE ir<%.447.us> = add ir<%.298.07.us>, ir<%.444.us>
    CLONE ir<%.448.us> = getelementptr ir<%arg.a.4>, ir<%.447.us>
    CLONE ir<%.449.us> = load ir<%.448.us>
    CLONE ir<%.503.us> = getelementptr ir<%arg.out.4>, ir<%.447.us>
    CLONE store ir<%.449.us>, ir<%.503.us>
    EMIT vp<%8> = VF * UF +(nuw)  vp<%1>
    EMIT branch-on-count  vp<%8> vp<%0>
  No successors
}
No successors
}
VPlan 'Initial VPlan for VF={2},UF>=1' {
Live-in vp<%0> = vector-trip-count

<x1> vector loop: {
  B52.us:
    EMIT vp<%1> = CANONICAL-INDUCTION
    WIDEN-INDUCTION %.298.07.us = phi 0, %.349.us
    CLONE ir<%.447.us> = add ir<%.298.07.us>, ir<%.444.us>
    CLONE ir<%.448.us> = getelementptr ir<%arg.a.4>, ir<%.447.us>
    WIDEN ir<%.449.us> = load ir<%.448.us>
    CLONE ir<%.503.us> = getelementptr ir<%arg.out.4>, ir<%.447.us>
    WIDEN store ir<%.503.us>, ir<%.449.us>
    EMIT vp<%8> = VF * UF +(nuw)  vp<%1>
    EMIT branch-on-count  vp<%8> vp<%0>
  No successors
}
No successors
}
LV: Found an estimated cost of 0 for VF 1 For instruction:   %.298.07.us = phi i64 [ 0, %B40.endif.us ], [ %.349.us, %B52.us ]
LV: Found an estimated cost of 1 for VF 1 For instruction:   %.349.us = add nuw nsw i64 %.298.07.us, 1
LV: Found an estimated cost of 1 for VF 1 For instruction:   %.447.us = add i64 %.298.07.us, %.444.us
LV: Found an estimated cost of 0 for VF 1 For instruction:   %.448.us = getelementptr double, double* %arg.a.4, i64 %.447.us
LV: Found an estimated cost of 1 for VF 1 For instruction:   %.449.us = load double, double* %.448.us, align 8
LV: Found an estimated cost of 0 for VF 1 For instruction:   %.503.us = getelementptr double, double* %arg.out.4, i64 %.447.us
LV: Found an estimated cost of 1 for VF 1 For instruction:   store double %.449.us, double* %.503.us, align 8
LV: Found an estimated cost of 1 for VF 1 For instruction:   %exitcond.not = icmp eq i64 %.349.us, %arg.out.5.1
LV: Found an estimated cost of 0 for VF 1 For instruction:   br i1 %exitcond.not, label %B50.B38.loopexit_crit_edge.us, label %B52.us
LV: Scalar loop costs: 5.
LV: Found an estimated cost of 0 for VF 2 For instruction:   %.298.07.us = phi i64 [ 0, %B40.endif.us ], [ %.349.us, %B52.us ]
LV: Found an estimated cost of 1 for VF 2 For instruction:   %.349.us = add nuw nsw i64 %.298.07.us, 1
LV: Found an estimated cost of 1 for VF 2 For instruction:   %.447.us = add i64 %.298.07.us, %.444.us
LV: Found an estimated cost of 0 for VF 2 For instruction:   %.448.us = getelementptr double, double* %arg.a.4, i64 %.447.us
LV: Found an estimated cost of 1 for VF 2 For instruction:   %.449.us = load double, double* %.448.us, align 8
LV: Found an estimated cost of 0 for VF 2 For instruction:   %.503.us = getelementptr double, double* %arg.out.4, i64 %.447.us
LV: Found an estimated cost of 1 for VF 2 For instruction:   store double %.449.us, double* %.503.us, align 8
LV: Found an estimated cost of 1 for VF 2 For instruction:   %exitcond.not = icmp eq i64 %.349.us, %arg.out.5.1
LV: Found an estimated cost of 0 for VF 2 For instruction:   br i1 %exitcond.not, label %B50.B38.loopexit_crit_edge.us, label %B52.us
LV: Vector loop of width 2 costs: 2.
LV: Selecting VF: 2.
LV(REG): Calculating max register usage:
LV(REG): At #0 Interval # 0
LV(REG): At #1 Interval # 1
LV(REG): At #2 Interval # 2
LV(REG): At #3 Interval # 2
LV(REG): At #4 Interval # 3
LV(REG): At #5 Interval # 3
LV(REG): At #7 Interval # 1
LV(REG): VF = 2
LV(REG): Found max usage: 2 item
LV(REG): RegisterClass: Generic::ScalarRC, 3 registers
LV(REG): RegisterClass: Generic::VectorRC, 1 registers
LV(REG): Found invariant usage: 1 item
LV(REG): RegisterClass: Generic::VectorRC, 1 registers
LV: The target has 16 registers of Generic::ScalarRC register class
LV: The target has 16 registers of Generic::VectorRC register class
LV: Loop cost is 5
LV: IC is 2
LV: VF is 2
LV: Interleaving to reduce branch cost.
LV: Found a vectorizable loop (2) in foo
LV: Interleave Count is 2
LEV: Epilogue vectorization is not profitable for this loop
Executing best plan with VF=2, UF=2
LV: Interleaving disabled by the pass manager
```

**Comment by sklam** [comment]
Actually, it's fixed on main https://github.com/numba/numba/commit/59adb07109d75745a93591fb77513d5f5c44c8ae
